### PR TITLE
Port Difference & Sym.Diff. to C++, cleanups and tests

### PR DESCRIPTION
**Pull request by wonder-sk**
Following up on the work started in #6811 

A major cleanup is that symmetrical difference is now implemented as `(A - B) U (B - A)` and the code for `A - B` is shared between "difference" and "symmetrical difference" algorithms.

Next step: union algorithm. The existing python code should be replaced by a combination of overlay algorithms we already have: `(A intersection B) U (A - B) U (B - A)`

### CODE REVIEW

File: src/analysis/processing/qgsoverlayutils.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      }
+
+      QgsFeature outFeat;
+      outFeat.setGeometry( geom );
+      outFeat.setAttributes( attrs );
+      sink.addFeature( outFeat, QgsFeatureSink::FastInsert );
+    }
+    else
+    {
+      // TODO: should we write out features that do not have geometry?
```
Comment (nyalldawson) [inline src/analysis/processing/qgsoverlayutils.cpp]: I think so, yes

File: src/analysis/processing/qgsoverlayutils.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+          break;
+
+        if ( engine->intersects( featB.geometry().constGet() ) )
+          geometriesB << featB.geometry();
+      }
+
+      if ( !geometriesB.isEmpty() )
+      {
+        QgsGeometry geomB = QgsGeometry::unaryUnion( geometriesB );
+        geom = geom.difference( geomB );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsoverlayutils.cpp]: Do we need a wkb type check here? E.g. if input layer is polygons, throw away points/lines (is that even possible? Maybe not...). But definitely we need multipart handling at least - e.g. polygon layer which results in multipolygon results (split to multiple features? Abort with error?)

**Review by nyalldawson — COMMENTED** [review]
Awesome! One thing I'm wondering - should the difference alg be a QgsFeatureBasedAlgorithm? It's looping through features in sourceA independently, so I think it's a good candidate.

The benefit would be when https://github.com/qgis/QGIS-Enhancement-Proposals/issues/114 is implemented - then we would have any easy way to chop out parts of geometries overlapped by another layer in place. This is one of the few remaining tasks which is a nightmare to do in qgis, yet a few simple clicks in MapInfo...

**Comment by wonder-sk** [comment]
> should the difference alg be a QgsFeatureBasedAlgorithm?

I would prefer not to make it feature-based alg. Partially for consistency with other algs ("why union and sym.difference are not feature-based?"), but mainly because my plan is to sort out also the case of self-intersecting input features (probably as an option) and for doing that it would be good to have extra pre- or post-processing code to resolve self-intersections - and it won't be a single-pass alg anymore.

File: src/analysis/processing/qgsoverlayutils.cpp (hunk 2 above)
Comment (wonder-sk) [reply src/analysis/processing/qgsoverlayutils.cpp]: > Do we need a wkb type check here?

Compared to intersection, we can't get a mixed bag of geometries from difference (AFAIK). We can only erase the whole geometry by subtracting "geomB" that completely covers "geom" - a case that we should actually handle properly and skip such feature (right now it is skipped with an error)

> But definitely we need multipart handling at least

This is already handled just like in the intersection algorithm - the sink is created as a multi-part geometry type. Test data produce both polygons and multi-polygons. Maybe we could add `geom.convertToMultiType()` to be safe with stricter data providers?

SAGA has an option to split multi-part geometries on output into single-part geometries. We could have such option at some point too.

File: src/analysis/processing/qgsoverlayutils.cpp (hunk 1 above)
Comment (wonder-sk) [reply src/analysis/processing/qgsoverlayutils.cpp]: Cool... will check also how other engines handle that, and then lock down the behavior to be consistent across algs with another piece of test data with no geometries. Would leave the to-do there as a reminder if you don't mind :-)

File: src/analysis/processing/qgsoverlayutils.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/analysis/processing/qgsoverlayutils.cpp]: Ok, +1 to force multipart via convertToMultipart then

**Comment by nyalldawson** [comment]
I guess we can always add a second algorithm which is feature based and hidden from the toolbox outside of the feature based mode...

File: src/analysis/processing/qgsoverlayutils.cpp (hunk 1 above)
Comment (agiudiceandrea) [reply src/analysis/processing/qgsoverlayutils.cpp:114]: Hi @wonder-sk, have you had the chance to check? Could you have a look at https://github.com/qgis/QGIS/issues/61205 which I guess it is related.

Comment (wonder-sk) [reply src/analysis/processing/qgsoverlayutils.cpp:114]: hi @agiudiceandrea I believe this should be unrelated - this specific to-do was about whether features without geometry should get written - but the mentioned bug report does not have features without geometry...

Comment (agiudiceandrea) [reply src/analysis/processing/qgsoverlayutils.cpp:114]: Probably I misunderstood the issue report and/or the TODO, anyway it seems to me the issue is actually about the fact that the output of the Difference algorithm incorrectly contains a feature without geometry.

Comment (wonder-sk) [reply src/analysis/processing/qgsoverlayutils.cpp:114]: I see... this branch with the TODO is for the case when the _input_ feature has null geometry. From what I understand, in that bug report all input features have non-null geometry, so the issue must be elsewhere...

**Pull request merged**